### ISSUE TITLE
[BUG] Node driver provisioning fails when resource quota configured in project

### ISSUE THREAD

**Issue by janeczku** [harvester/harvester#1455]
Harvester 0.3.0 / Rancher 2.6.1

**Scenario 1:**

1. Create a new project in Rancher Virtualization Management and configure resource quota for CPU reservation, CPU limit and memory reservation, memory limit. 
2. Create a namespace in the project
3. Provision a cluster in the namespace using Harvester RKE2 node driver. 

**Result:**

VM fails to be scheduled with error:

```
8m19s       Warning   FailedCreate            virtualmachineinstance/test321-pool1-e80882ca-mm4ch        
Error creating pod: pods "virt-launcher-test321-pool1-e80882ca-mm4ch-g59vt" is forbidden:
failed quota: default-drbgs: must specify limits.memory
```

**Scenario 2:**

1. Create a new project in Rancher Virtualization Management and configure a resource quota for CPU and memory (requests and limits)
2. Create a namespace in the project and configure default "Container Resource Limit" with:
```
CPU limit/reservation: 2000/1000
Memory limit/reservation: 500/256
```
3. Provision a cluster in the namespace using Harvester RKE2 node driver and configuring 1GB of memory for the node pool.

**Result:**

VM fails to be scheduled with error:

```
FailedCreate | VirtualMachine test123-test123-b814f7b6-f6tf9Error creating virtual machine instance: 
admission webhook "virtualmachineinstances-create-validator.kubevirt.io" denied the request: 
spec.domain.resources.requests.memory '1Gi' is greater than spec.domain.resources.limits.memory '256Mi'
```

**Comment by weihanglo** [comment]
For "Scenario 1", the root cause is that the VMBUilder doesn't specify `CPU resources.limits`[^1].

For "Scenario 2", I have no idea how memory `resources.limits` was set to 256Mi, but it seems reasonable for me to reject the provision since the VM (1GiB) exceeds the memory quota (500Mi). Maybe I reproduced it wrong. Could you detail the steps? Thanks.

Anyhow, both scenario seem could be be resolved by simply specifying the correct resource  request/limit. In addition, we're going to accept resource overcommit feature[^2]. Should node driver support overcommitment or just provision as is (limit = request)?


[^1]: https://github.com/harvester/harvester/blob/119c2c484efb9dbb75333cd0a88bc2d67c42e3d1/pkg/builder/vm.go#L148-L151
[^2]: https://github.com/harvester/harvester/issues/1429

**Comment by weihanglo** [comment]
Also note that KubeVirt requests additional memory[^1] more than we specify for a VM, so we might meet a scenario that we cannot provision a 4GiB VM within a 4GiB quota namespace. Surely we can turn off the guest overhead, but should we?

[^1]: https://kubevirt.io/user-guide/operations/node_overcommit/#overcommit-the-guest-overhead

**Comment by weihanglo** [comment]
This should be resolved by harvester/docker-machine-driver-harvester#27.

~~For whom aim to test this bugfix, please head to the node-driver page in Rancher. Then change to URL of harvester node dirver to https://releases.rancher.com/harvester-node-driver/v0.3.3/docker-machine-driver-harvester-amd64.tar.gz . You might want to remove the checksum to skip the validation.~~

Please use v0.3.4
https://github.com/harvester/harvester/issues/1455#issuecomment-972812490

**Comment by futuretea** [comment]
docker-machine-driver-harvester v0.3.3
harvester master version 45e2593
rancher 2.6.2
create Harvester cluster rke1/rke2 by node driver:
![image](https://user-images.githubusercontent.com/15064560/142408227-b28ea867-9e51-46fc-bf4a-e25086c78aff.png)
![image](https://user-images.githubusercontent.com/15064560/142408239-e6b7f6f3-ecef-4e28-8360-eb42536b50c5.png)

**Comment by weihanglo** [comment]
> per-device boot elements cannot be used together with os/boot elements

Looks like a device boot order issue. Setting hotplug flag to false might help? I am not sure. Need some tests.

https://github.com/harvester/docker-machine-driver-harvester/blob/b6ec9b6b5ab5dfbb6fda8f9b872bf2c33e874360/harvester/create.go#L110

**Comment by futuretea** [comment]
hotplug flag need to be false, tested

**Comment by futuretea** [comment]
@weihanglo  created a new release v0.3.4  for test: https://releases.rancher.com/harvester-node-driver/v0.3.4/docker-machine-driver-harvester-amd64.tar.gz

**Comment by TachunLin** [comment]
Verified **not fixed** on `master-d577bb6c-head`. with harvester node driver `v0.3.4`
The first scenario and the second both failed in `Failing bootstrap machines` 
Please refer to the test result. Move it back to implement pipeline

### Result
For the **first** scenario of resource quota configured on project side only (not in namespace), 
Provision harvester RKE2 node driver by configuring 2GB of memory failed to create virtual machine with error

`Failed to bootstrap machine ... failed creating server`

![image](https://user-images.githubusercontent.com/29251855/142760217-c66bdc0c-ac0b-4eb1-955d-cb9815993eb6.png)

![image](https://user-images.githubusercontent.com/29251855/142760255-47619847-ca34-4abd-b8c8-4e5ffa3d9355.png)


For the **second** scenario of resource quota configured on both project and namespace side. 
Provision harvester RKE2 node driver by configuring 1GB of memory failed to create virtual machine with error 
`Failed to bootstrap machine ... failed creating server`

![image](https://user-images.githubusercontent.com/29251855/142759559-c51ca5dc-40ec-4546-a227-07be9ebbf580.png)

![image](https://user-images.githubusercontent.com/29251855/142759580-99eb55f8-13b6-40a0-857c-03bea3ce4e6d.png)

### Test Information
* Test Environment: 3 harvester nodes on local kvm machines                    
* Harvester version: master-d577bb6c-head
* Node driver version: [v0.3.4](https://releases.rancher.com/harvester-node-driver/v0.3.4/docker-machine-driver-harvester-amd64.tar.gz)
* Rancher version: 2.6.2

### Verify Steps

1. Import harvester from rancher 
2. Open Cluster management -> Driver
3. Find and edit harvester node driver 
4. Change the url to v0.3.4  
https://releases.rancher.com/harvester-node-driver/v0.3.4/docker-machine-driver-harvester-amd64.tar.gz 
![image](https://user-images.githubusercontent.com/29251855/142759066-9fa5bbd4-803d-41f2-b4cc-4b2b9244353a.png)
5. Create cloud credential
6. Access harvester from virtualization management page
7. Create a new project `test-project` and assign the default resource quota value for cpu limit, cpu reseration, memory limit and memory reservation
![image](https://user-images.githubusercontent.com/29251855/142611905-fdec3bd3-5905-4684-8ee6-30238ec54313.png)
8. Create a namespace `test-namespace` under the `test-project`
8. Provision a rke2 cluster with 2 CPU and 2GB memory and set namespace to `test-namespace1`
![image](https://user-images.githubusercontent.com/29251855/142758187-b4f38094-caac-4b99-ac4f-6c64ab480b84.png)

9. Create a new project `test-project2` and assign the default resource quota value for cpu limit, cpu reseration, memory limit and memory reservation
![image](https://user-images.githubusercontent.com/29251855/142626526-22cff71b-57a0-421f-9a85-947af333827e.png)
10. Create a namespace `test-namespace2` under the `test-project2`and set container resource quota
![image](https://user-images.githubusercontent.com/29251855/142626687-e94a595e-90e4-4f2a-9e4d-8e6b1049b218.png)

11. Provision a rke2 cluster with 2 CPU and 1GB memory and set namespace to `test-namespace2`

![image](https://user-images.githubusercontent.com/29251855/142759153-ffe3e3eb-7a02-4e6e-bb83-e9b919c60b4a.png)

**Comment by futuretea** [comment]
The resourcequotas created from Harvester's UI don't match the actual units on the page.
![image](https://user-images.githubusercontent.com/15064560/142763651-a3ac7d4b-d485-43b8-815c-a52150ef9531.png)
![image](https://user-images.githubusercontent.com/15064560/142763711-a6c70643-fe33-4964-8cea-b472e946ec5c.png)
@WuJun2016

**Comment by WuJun2016** [comment]
Fixed in rancher: https://github.com/rancher/dashboard/issues/4417

**Comment by futuretea** [comment]
@TachunLin you can try to use [rancher v2.6.3-rc1](https://github.com/rancher/rancher/releases/tag/v2.6.3-rc1)

**Comment by futuretea** [comment]
@TachunLin sorry, need waiting https://github.com/rancher/rancher/pull/35597

**Comment by futuretea** [comment]
@TachunLin If using rancher v2.6.3-rc1,  since harvester had been changed to build-in, we can only use kubectl （rancher local cluster）to update node driver to v0.3.4

```bash
 kubectl edit nodedriver harvester
```

- change spec.checksum to ""
- change spec.buildin to false
- change spec.url to "https://releases.rancher.com/harvester-node-driver/v0.3.4/docker-machine-driver-harvester-amd64.tar.gz"

![image](https://user-images.githubusercontent.com/15064560/143041870-9f922baa-6d00-454f-8e2f-868c47d239e0.png)

**Comment by TachunLin** [comment]
Hi @futuretea , thanks for the confirmation and the instruction to update node driver to v0.3.4.

Before I knew you solution, I found a different way to update node driver to v0.3.4 
By using `View in API` function, I am able to edit , update url, remove the built-in and save
![image](https://user-images.githubusercontent.com/29251855/143049017-44a78a05-7a92-45eb-9a84-815d49bf0d0e.png)

Today I tried using rancher v2.6.3-rc1 to verify this issue, but still failed in the first scenario 
I think we need  move it to implement and wait for [rancher/rancher#35597 ](https://github.com/rancher/rancher/pull/35597) merged to verify again

Here is the verification steps for your reference.

By setting enough resource quota on new project 

- cpu limit/reservation: 6 core / 6 core
- memory limit/reservation: 6GB / 6GB

![image](https://user-images.githubusercontent.com/29251855/143021890-6821750c-7bae-40d6-882e-2018a93a745c.png)

And provision a 4 core / 4 GB RKE2 cluster which within the range of resource allocation
![image](https://user-images.githubusercontent.com/29251855/143021979-e95a58f4-2bc8-469f-8d7b-cf4519576048.png)

RKE2 vm failed to provision on harvester with incompatible resource request
```
Error creating pod: pods "virt-launcher-rke2-sc1-pool1-bc3309e7-g57mp-fx6qz" is forbidden: exceeded quota: default-6dhmn, requested: limits.memory=4503982081, used: limits.memory=0, limited: limits.memory=4Gi
```
![image](https://user-images.githubusercontent.com/29251855/143022277-f52aa0d1-bc4b-4e31-83d9-9678db1a5a68.png)

**Comment by futuretea** [comment]
@TachunLin  According to the error:
![image](https://user-images.githubusercontent.com/15064560/143056293-72b1d476-d79d-4dfd-99f1-cddde9d8f59a.png)
the actual resourcequota limit is 4Gi,
Did you set the display to be 4Gi at the beginning, and then changed to 6Gi?
I find that editing the resourcequota  configuration on the UI doesn't seem to work.

**Comment by TachunLin** [comment]
@futuretea, in this scenario, I create a brand new project and set resource quota to 6Gi in the beginning without edit it afterward.
I remember also tried changing from 4Gi to 6Gi, the actually quota still set to 4Gi on harvester VM event page.
I will try again and file issue if needed

**Comment by guangbochen** [comment]
@TachunLin we should be able to validate it using the `master-head` version of the Rancher server, can u please re-validate it again, thanks.

**Comment by TachunLin** [comment]
Verified **fixed** on Rancher `v2.6-head` (4375161) with harvester node driver `v0.3.4` (builtin). Close this issue. 


### Result
The `1st scenario` can provision RKE2 cluster with VM successfully created and running with resource quota configured on project and default setting on namespace.  (No error)

![image](https://user-images.githubusercontent.com/29251855/143866315-223652fe-8b21-4293-a30a-d87c9f5c40ca.png)

Successfully created (under resource limit) 
![image](https://user-images.githubusercontent.com/29251855/143866706-5107e4df-da7d-41b0-8756-b3561f8d49c2.png)

Successfully provision RKE2 cluster
![image](https://user-images.githubusercontent.com/29251855/143867789-0349c6dc-cbc4-4284-a28c-2795b1499549.png)

The `2nd scenario` can provision RKE2 cluster with VM successfully created and running with resource quota configured on project, default setting on namespace, and container resource quota configured (No error) 

![image](https://user-images.githubusercontent.com/29251855/143870735-286b0e00-b141-4028-85cb-137d9933d51f.png)

![image](https://user-images.githubusercontent.com/29251855/143871373-dbffdf82-e60a-4723-8301-67d7fa8225a4.png)


### Test Information
* Test Environment: 3 harvester nodes on local kvm machines                    
* Harvester version: master-47ea94ee-head
* Node driver version: [v0.3.4](https://releases.rancher.com/harvester-node-driver/v0.3.4/docker-machine-driver-harvester-amd64.tar.gz)
* Rancher version: `2.6-head`

### Verify Steps

1. Import harvester from rancher 
2. Open Cluster management -> Driver
3. Create cloud credential
4. Access harvester from virtualization management page
5. Create a new project `test-1561-proj` and assign the following resource quota value

 * Project resource quota: 
   - CPU Limit / CPU Reservation: 6000 / 6144
   - Memory Limit / Memory Reservation: 6000 / 6144
 
 * Namespace resource quota: None (default) 

6. Create a namespace `test-1561-ns` under the `test-1561-proj`
7. Provision a rke2 cluster with 2 CPU and 4GB memory and set namespace to `test-1561-ns`
8. Create a new project `test-project2` and and assign the following resource quota value
    
* Project resource quota: 
  - CPU Limit / CPU Reservation: 6 core / 6GB
  - Memory Limit / Memory Reservation: 6 core / 6GB
 
 * Namespace resource quota: None (default)
 * Container resource limit:
   - CPU limit/reservation: 2000/1000
   - Memory limit/reservation: 500/256

![image](https://user-images.githubusercontent.com/29251855/143868981-ba38d935-63df-468a-a6d8-58189abc4671.png)

11. Provision a rke2 cluster with `1` CPU and `1GB` memory and set namespace to `test-namespace2`
![image](https://user-images.githubusercontent.com/29251855/143870073-2eb56d39-ae21-4b50-ba3f-b035d258bb2d.png)
